pkg/auth/handler/ssohandler/auth_handler.go

```diff
@@ -0,0 +1,398 @@
+		Dependency: authDependency,
+	}).Methods("OPTIONS", "GET")
+	return server
+}
+
+type AuthHandlerFactory struct {
+	Dependency auth.DependencyMap
+}
+
+func (f AuthHandlerFactory) NewHandler(request *http.Request) http.Handler {
+	h := &AuthHandler{}
+	inject.DefaultInject(h, f.Dependency, request)
+	vars := mux.Vars(request)
+	h.ProviderName = vars["provider"]
+	// since auth_hander need create different responses depends on ux_mode,
+	// so here has a APIHandler to handle those different situation.
+	return h.Handler()
+}
+
+func (f AuthHandlerFactory) ProvideAuthzPolicy() authz.Policy {
+	return policy.AllOf()
+}
+
+// AuthRequestPayload login handler request payload
+type AuthRequestPayload struct {
+	Code         string
+	Scope        sso.Scope
+	EncodedState string
+}
+
+// Validate request payload
+func (p AuthRequestPayload) Validate() error {
+	if p.Code == "" {
+		return skyerr.NewInvalidArgument("Authorization Code is required", []string{"code"})
+	}
+
+	if p.EncodedState == "" {
+		return skyerr.NewInvalidArgument("EncodedState is required", []string{"state"})
+	}
+
+	return nil
+}
+
+// AuthHandler decodes code response and fetch access token from provider.
+//
+// curl http://localhost:3000/sso/<provider>/auth_handler?code=<code>&state=<state>
+//
+// For ux_mode is 'ios' or 'android',
+// it creates a 302 response, and Location points to:
+// myapp://user.skygear.io/sso/{provider}/auth_handler?result=
+//
+// Fox ux_mode is 'web_redirect',
+// it creates a 302 response, and Location points to: sso_callback_url
+// and set cookie in the response.
+//
+// For ux_mode is 'web_popup',
+// it will render a html page and set cookie in the response.
+//
+type AuthHandler struct {
+	TxContext               db.TxContext                `dependency:"TxContext"`
+	AuthContext             coreAuth.ContextGetter      `dependency:"AuthContextGetter"`
+	Provider                sso.Provider                `dependency:"SSOProvider"`
+	OAuthAuthProvider       oauth.Provider              `dependency:"OAuthAuthProvider"`
+	AuthInfoStore           authinfo.Store              `dependency:"AuthInfoStore"`
+	RoleStore               role.Store                  `dependency:"RoleStore"`
+	TokenStore              authtoken.Store             `dependency:"TokenStore"`
+	AuthHandlerHTMLProvider sso.AuthHandlerHTMLProvider `dependency:"AuthHandlerHTMLProvider"`
+	ProviderName            string
+}
+
+func (h AuthHandler) WithTx() bool {
+	return true
+}
+
+func (h AuthHandler) DecodeRequest(request *http.Request) (handler.RequestPayload, error) {
+	payload := AuthRequestPayload{}
+	q := request.URL.Query()
+	payload.Code = q.Get("code")
+	payload.Scope = strings.Split(q.Get("scope"), " ")
+	payload.EncodedState = q.Get("state")
+
+	return payload, nil
+}
+
+func (h AuthHandler) Handler() http.Handler {
+	// reference from APIHandlerToHandler
+	return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
```

Comment:
Since we are not using APIHandler in this handler, we don't need to have a `Handle` function to return `HandlerFunc`. `AuthHandler` can be the handle function directly, see sso `IFrameHandler`.